Monitor Conditions

simshaun opened the pull request:
Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Hello,

We have a need at work for monitoring values of DNS records. I got (really) bored and decided to do something about it. I understand this is a semi-major feature and may need some changes, or maybe you don't like it all. That's fine.

This PR introduces support for setting up conditions that affect UP/DOWN status on monitors. It pertains to #3919 and eventually, I think, could obviate the need for the keyword type, and maybe the JSON Query type.

It is opt-in per type, and I've only opted-in the DNS type for now. So, this doesn't break existing keyword/JSON query types. It's generic but extendable enough in my opinion that it could allow implementing a wide variety of scenarios without needing further UI changes going forward.

Currently missing the ability to drag conditions around, but that shouldn't be too difficult to shoehorn in eventually.

Screenshots below.


## Type of change

- User interface (UI)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [x] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

![image](https://github.com/user-attachments/assets/0da1c507-3d2f-4a13-a518-2dd1b98bf9c3)

Groups:

![image](https://github.com/user-attachments/assets/1dedbe1b-a641-44a9-9c91-6c67168d5d3b)

CommanderStorm commented on db/knex_migrations/2024-08-24-0000-conditions.js:
```diff
@@ -0,0 +1,12 @@
+exports.up = function (knex) {
+    return knex.schema
+        .alterTable("monitor", function (table) {
+            table.text("conditions").defaultTo(null);
```
While we have a similar filed already (See https://github.com/louislam/uptime-kuma/pull/4717/files), this is likely a better implementation.
We can delete the other column when we refactor said Monitor or when jsonata support is added to the Operators..

simshaun commented:
lint issues are because I'm using a few imports for JSDoc type hints. I see the ESLint rule that allows this is disabled: `"jsdoc/no-undefined-types": "off"`.

I can remove the imports that enable navigation in IDE or I could enable that ESLint rule.

CommanderStorm reviewed (commented):
Looks very nice. Good job.
The testcases are most appreciated ❤️

I have left a few inline comments below. ^^

simshaun commented on server/monitor-conditions/operators.js:
```diff
@@ -0,0 +1,201 @@
+     * @inheritdoc
+     */
+    test(variable, value) {
+        // eslint-disable-next-line eqeqeq
+        return variable == value;
```
I initially was going to add an `mx_priority` variable to check MX record priorities. Node's DNS module parses MX records and returns priority as a number type.

In this new conditions UI, comparison values are stored as strings. i.e. My condition would be stored in the monitor as `mx_priority equals "10"`.

I have a set of numeric operators (less than, greater than, etc) that cast the value to Number internally, but the EqualsOperator is currently used for both strings and numbers. It needs loose equality to handle 10 == "10".

An alternate solution could be to rename this operator to StringEquals (make it strict) and add a new NumberEquals operator (also strict) that casts value internally.

CommanderStorm replied:
Adding a comment that this is because of sharing between numbers and strings to handle `10 == "10"` is likely fine.

simshaun replied:
I'm flip-flopping on this. I think it may be advantageous to start out with strict operators rather than maybe trying to implement them later on (and have to worry about backwards compatibility.)

So, require a monitor type to explicitly use either `StringEquals` or `NumberEquals` when declaring its supported condition variables.

simshaun commented on server/monitor-types/dns.js:
```diff
@@ -1,12 +1,22 @@
+    supportsConditions = true;
+
+    conditionVariables = [
+        new ConditionVariable("record", defaultStringOperators ),
```
Possible, but would need to carefully migrate existing monitors or retain backwards-compatible code. I think this is out-of-scope of an initial implementation and while I agree with the idea, should probably "kick the can" down the road.

simshaun commented on src/components/EditMonitorConditions.vue:
```diff
@@ -0,0 +1,143 @@
+            </template>
+        </div>
+        <div class="monitor-conditions__buttons">
+            <button class="btn btn-outline-secondary btn-sm me-2" type="button" data-testid="add-condition-button" @click="addCondition">
```
![image](https://github.com/user-attachments/assets/8b1eba94-29e5-4eed-8a18-525909ae4539)

I have a concern that this will get accidentally tapped by mobile users.

Could space the buttons out more; Or add a confirmation dialog (annoying imo); Or keep it above the fields.

simshaun replied:
I'm experimenting with keeping the variable select aligned with the next rows but it only looks good visually if the next row is a condition. When the next row is a group, it gets whacky. IMO better to just leave it alone.

simshaun replied:
> maybe nicer if aligned

Experimenting with this and it's janky to implement.

simshaun replied:
> X on "Add Condition" button

Not sure what you mean here. Only have Groups at the root level?

simshaun replied:
![image](https://github.com/user-attachments/assets/79302326-2e3e-4aad-aad1-858036a3b327)

Is this necessary? A group is a group. I don't think it needs to be "Add Condition Group".

CommanderStorm replied:
> Not sure what you mean here. Only have Groups at the root level

I was thinking about https://github.com/louislam/uptime-kuma/pull/5048#discussion_r1730428445

=> lets move that subpart there, sorry

CommanderStorm replied:
> Could space the buttons out more;

Lets do that. And maybe make the delete buttons' forground `text-color` red if you think this accidental tapping is a concearn.

simshaun replied:
Updated to:

![image](https://github.com/user-attachments/assets/05fc583a-fef2-451c-a06a-ccbafe45dd7a)

simshaun commented on src/components/EditMonitorConditions.vue:
```diff
@@ -0,0 +1,143 @@
+    </div>
+</template>
+
+<script>
```
I’m not sure that’s the best approach, as a condition may not be necessary for most monitors. (i.e. They are optional.)

CommanderStorm replied:
I don't think we have a monitor that does not do a check.
`dns` was kind of the odd ones out in that way which just did a DNS lookup without really looking at the output..

=> having a default group be added in the frontend (not the migration for compatability) likely makes sense..

simshaun commented on server/monitor-types/dns.js:
```diff
@@ -17,36 +27,65 @@ class DnsMonitorType extends MonitorType {
+                : conditionsResult;
         } else if (monitor.dns_resolve_type === "CNAME" || monitor.dns_resolve_type === "PTR") {
             dnsMessage += dnsRes[0];
+            conditionsResult = conditions
```
I've revamped this file. Take a look.

simshaun commented on server/monitor-types/monitor-type.js:
```diff
@@ -1,6 +1,21 @@
+const { ConditionVariable } = require("../monitor-conditions/variables");
```
Changed to imports within JSDoc

simshaun commented on test/backend-test/monitor-conditions/test-operators.js:
```diff
@@ -0,0 +1,88 @@
+
+test("Test EqualsOperator", async (t) => {
+    const op = operatorMap.get(OP_EQUALS);
+    assert.strictEqual(true, op.test("foo", "foo"));
```
Yea, not a fan of `assert.ok`. There's no mental leaps required when reading `assert.strictEqual` but `assert.ok` always makes me wonder "Well what does Node consider 'ok'?"

simshaun commented on src/components/EditMonitorCondition.vue:
```diff
@@ -0,0 +1,138 @@
+            </option>
+        </select>
+
+        <input v-model="model.value" type="text" class="form-control" :aria-label="$t('Value')" data-testid="condition-value">
```
I've revamped this and several other strings to system keys in the language file, like:

```json
"conditionAdd": "Add Condition",
"conditionDelete": "Delete Condition",
"conditionAddGroup": "Add Group",
"conditionDeleteGroup": "Delete Group",
"conditionValuePlaceholder": "Value",
```

simshaun commented on src/components/EditMonitorCondition.vue:
```diff
@@ -0,0 +1,138 @@
+    margin-left: auto;
+}
+
+@container (min-width: 650px) {
```
Updated

simshaun commented on src/components/EditMonitorCondition.vue:
```diff
@@ -0,0 +1,138 @@
+            </option>
+        </select>
+
+        <input v-model="model.value" type="text" class="form-control" :aria-label="$t('Value')" data-testid="condition-value">
```
Probably.

Maybe not if there were a "not blank" operator, but I haven't implemented that. Would just hide the value field in that case anyway.

simshaun commented:
Rebased to master

CommanderStorm commented on server/monitor-types/dns.js:
```diff
@@ -17,36 +27,56 @@ class DnsMonitorType extends MonitorType {
-            dnsMessage = dnsMessage.slice(0, -2);
+        const conditions = ConditionExpressionGroup.fromMonitor(monitor);
+        let conditionsResult = null;
+        const handleConditions = (data) => conditions ? evaluateExpressionGroup(conditions, data) : null;
```
Follow-up to https://github.com/louislam/uptime-kuma/pull/5048#discussion_r1730408556

Can `conditions` ever be falsy?

I don't think so, right? => Am I missing something?

CommanderStorm replied:
Followup to https://github.com/louislam/uptime-kuma/pull/5048#discussion_r1730408556

An empty expression group should evaluate to true always.
That simplifies the code and removes the `null`

simshaun commented:
Since #5056 was merged, I've fixed, added to, and improved the E2E tests for this PR.

I'll go through your remaining comments sometime later today.

simshaun commented:
I think I've handled all the remaining issues.

CommanderStorm commented:
Thanks for the enhancement! 🎉 

<img width="200px" src="https://media4.giphy.com/media/3oEdva9BUHPIs2SkGk/giphy.gif"/>

> [!NOTE]
> This PR is part of the [`v2.0` merge window](https://github.com/louislam/uptime-kuma/milestone/24) => see #4500 for the bugs that need to be addressed before we can ship this release ^^

<sub>All changes in this PR are small and uncontroversial ⇒ merging with junior maintainer approval</sub>

vishalsabhaya commented:
@simshaun @CommanderStorm 

i am working on below improvement.
https://github.com/louislam/uptime-kuma/pull/5025
i want to do final testing so i taken pull from master branch but i am getting below error.
any idea? why
<img width="1103" alt="image" src="https://github.com/user-attachments/assets/03d341b9-6348-4c3d-928d-4ccc44492f8b">

simshaun commented:
@vishalsabhaya 

The migration in this database changes sets a default value of `[]` for the conditions column in the database. I'm curious how yours is ending up empty.

vishalsabhaya commented:
@simshaun 

i am using mysql. is there any problem with MySQL? have you written migration for MySQL?

simshaun commented:
@vishalsabhaya The migration should be database agnostic.

https://github.com/louislam/uptime-kuma/pull/5048/files#diff-4111d2d321bfe6baf61dc1b78521a8ee9445bf86f599aac8fa3be00cb85139f1

Can you confirm via DBMS the default value of the conditions column in the monitor table?

louislam commented on db/knex_migrations/2024-08-24-0000-conditions.js:
```diff
@@ -0,0 +1,12 @@
+exports.up = function (knex) {
+    return knex.schema
+        .alterTable("monitor", function (table) {
+            table.text("conditions").notNullable().defaultTo("[]");
```
I am just testing it on MariaDB, and found a strange bug of this pr.

When I click login button, it doesn't work. In the console, it throws an exception: 

> Trace: SyntaxError: Unexpected end of JSON input
>    at JSON.parse (<anonymous>)
>  at Proxy.toJSON (C:\Users\Louis-K6\Dropbox\My Projects\uptime-kuma\server\model\monitor.js:159:30)

Line: 159
>        conditions: JSON.parse(this.conditions),


And then, I go into MariaDB and saw that is `No default`, while in my SQLite, it is `[]`. Any ideas?

![image](https://github.com/user-attachments/assets/ca677b57-220d-46be-87ac-051be41f9664)

louislam replied:
Omg... I just realized that `accepted_statuscodes_json` also have the same issue... 

Wait... What the hell: https://github.com/knex/knex/issues/2649

louislam replied:
Wanna cry, it don't have any issues using a raw sql..... 

```
ALTER TABLE `monitor`
	ADD COLUMN `condition_test` TEXT NOT NULL DEFAULT '[]' AFTER `rabbitmq_password`;
```

louislam replied:
Fixed: #5261
